### PR TITLE
Assignment5: @crawfordkid2 and @GeorgeKoralski007

### PR DESCRIPTION
**Pull request by GeorgeKoralski007**

### CODE REVIEW

File: GenericsHomework/GenericsHomework/Node.cs [hunk 1]
```diff
@@ -0,0 +1,54 @@
+﻿namespace GenericsHomework;
+
+public class Node<T>
+{
+    public T Value { get; set; }
```
Comment (jdriscoll1) [inline GenericsHomework/GenericsHomework/Node.cs:5]: Make this property nullable

File: GenericsHomework/GenericsHomework/Node.cs [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+            Node<T> next = current.Next;
+            current.Next = current;
+            current = next;
+        } while (current != this);
+    }
+
+    public void Append(T value)
+    {
+        if (Exists(value))
+            throw new ArgumentException("Value already exists in the list", nameof(value));
```
Comment (jdriscoll1) [inline GenericsHomework/GenericsHomework/Node.cs:33]: Looks great, although changing it to a format string such as: `ArgumentException($"Value already exists in the list {nameof(value)}")` or instead putting a space after "list" might make it neater.

File: GenericsHomework/GenericsHomeworkTests/NodeTests.cs [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+namespace GenericsHomework.Tests
+{
+    [TestClass]
+    public class NodeTests
+    {
+        [TestMethod]
+        public void Node_Constuctor_Success()
+        {
+            Node<int> node = new Node<int>(10);
```
Comment (jdriscoll1) [inline GenericsHomework/GenericsHomeworkTests/NodeTests.cs:12]: Change to `Node<int> node = new(10);`

**Review by jdriscoll1 — COMMENTED** [review]
Create a class library project called "GenericsHomework.". ✔
Create a node class that can contain a value of any type and points to the next node and traversing the next node points back to the first item. ✔
Define the Node class ✔
The values should be homogeneous. ✔ 
Include a constructor that takes a value. (No validation is necessary on the value). ✔
Add a ToString() override that writes out the value's ToString() result. ✔
Add a Next property that references the next node or else refers back to itself if there are no other nodes in the list. ✔
The Next property should be non-nullable (careful to follow the non-nullable property guidelines) ✔
The Next property setter should be private. ✔
Add an Append method that takes a value and appends a new Node instance after the current node (by invoking the Next property). ✔
Add a Clear method that effectively removes all items from a list except the current node. Pay attention as to whether you should be concerned with the following:
Whether it is sufficient to only set Next to itself ❌
Whether to set the removed items to circle back on themselves. In other words, whether to close the loop of the removed items. (Provide a test to show why this is required if it is required). ❌

Provide a comment & a test that shows why you chose to go through each node and set its next to itself
Given there is a circular list of items, provide a comment to indicate whether you need to worry about garbage collection because all the items point to each other and therefore may never be garbage collected. ❌
See above
Create an Exists method to test to see if a value exists in the list. ✔
Throw an error on an attempt to Append a duplicate value. (Make sure you test for this case) ✔
You should not rely on any BCL generic classes for your implementation. ✔
Extra Credit
Do one of the following two options (or both if you want extra, extra credit) :)
Implement a VennDiagram structure that contains n Circles that only contains homogenous reference types of any type. ❌
Each circle contains n items and each item can belong to one or more Circle instances.
You are not required to use a Node from earlier in the homework for your Venn diagram implementation.
You are welcome to use existing BCL generic classes for the extra credit.
Implement System.Collections.Generic.ICollection<T> on the Node class ❌
Fundamentals
Place all shared project properties into a Directory.Build.Props file.❌
Place all shared project items into a Directory.Build.targets file. ✔
nullable reference types is enabled ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ❌
Not enabled on GenericsHomeworkTests.cproj
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ❌
Lang version not set on tests
and enabled .NET analyzers for both projects ❌
EnableNETAnalyzers
For this assignment, always use Assert.AreEqual<T>() (the generic version) ❌
Add <type> on AreEqual tests in unit tests
All of the above should be unit tested ❌
Add unit test to display why you chose to do clear and you'll be good. 
Choose simplicity over complexity ✔

Overall it looks great, simple and clean design. I would recommend that you add comment and a test displaying why you chose to go through each value and close the loop in the clear method, and make a few tweaks.

File: GenericsHomework/GenericsHomework/Node.cs (hunk 1 above)
Comment (GeorgeKoralski007) [reply GenericsHomework/GenericsHomework/Node.cs:5]: Done

File: GenericsHomework/GenericsHomework/Node.cs (hunk 2 above)
Comment (GeorgeKoralski007) [reply GenericsHomework/GenericsHomework/Node.cs:33]: Added a space

File: GenericsHomework/GenericsHomeworkTests/NodeTests.cs (hunk 3 above)
Comment (GeorgeKoralski007) [reply GenericsHomework/GenericsHomeworkTests/NodeTests.cs]: Done

File: GenericsHomework/GenericsHomework/Node.cs [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+    public Node(T? value)
+    {
+        Value = value;
+        Next = this;
+    }
+
+    public override string ToString()
+    {
+        return Value?.ToString() ?? "null";
+    }
```
Comment (mmwoodfo) [inline GenericsHomework/GenericsHomework/Node.cs:33]:
```suggestion
    public override string? ToString()
    {
        return Value?.ToString();
    }

```
Could also do this

**Review by mmwoodfo — COMMENTED** [review]
Great job!

**Pull request closed without merging**